[EE-IR] Fragment Compiler Entrypoint in IR Backend

kandersen opened the pull request:
This commit introduces a new entrypoint for the IR backend, and starts
the work of accomoding the evaluator in psi2ir.

The evaluator expects a certain class and method structure of the
compiled fragment, but PSI is only supplied for the actual fragment.

So, to that end, this commit introduces a new "front end" of psi2ir
that "synthesizes" the module, class and method structure around the
fragment before calling into the existing psi2ir pipeline to obtain
the IR for the fragment.

The primary complication so far is handling the captured variables of
the fragment: they are mapped to parameters of the method surrounding
the fragment, and passed as arguments on evaluation. Hence, the IR
translation of the fragment needs to remap captured variables to the
appropriate parameter, in essentially all places they can be referred
to in the fragment (and hence in psi2ir). This commit introduces a
decorated symbol table that intercepts symbol look-ups and remaps as
appropriate.

Other cases that dispatches based on descriptor (see
`CallGenerator.kt`) needs other metadata to generate correct code.

It also introduces a shim in DeserializedContainerSource in the psi2ir
pipeline to facilitate facade class generation for the code that is
being debugged (which are generated as "external ir declarations").

Finally, in passing we resolve a small leftover from previous
refactoring that left an asssertion re. allowing IR to _assign_ to
parameters of methods.

kandersen commented:
This PR unblocks [PR #1666 in JetBrains/intellij-community/kt-212-master](https://github.com/JetBrains/intellij-community/pull/1666)

kandersen commented:
@udalov Sorry for the noise. I will let you know once I have addressed your initial comments from the previous version of these changes.

kandersen commented:
@udalov broke the dependency on the old backend. Please have a second look.

udalov commented on compiler/ir/ir.psi2ir/src/org/jetbrains/kotlin/psi2ir/generators/FragmentClassGenerator.kt:
```diff
@@ -0,0 +1,70 @@
+                context.symbolTable.nameProvider.nameForDeclaration(classDescriptor), visibility, modality
+            )
+        }.buildWithScope { irClass ->
+            irClass.thisReceiver = context.symbolTable.declareValueParameter(
```
Maybe just `irClass.createImplicitParameterDeclarationWithWrappedDescriptor()`

kandersen replied:
This does not appear to be in "scope" of psi2ir. Forgive me, but you must be referring to `compiler/ir/backend.common/src/org/jetbrains/kotlin/backend/common/ir/IrUtils.kt:createImplicitParameterDeclarationWithWrappedDescriptor`?

 As far as I can tell, that resides in `:compiler:backend-common`, already imported by psi2ir.

I must just be holding it wrong :sweat_smile:

udalov replied:
Oh, indeed, sorry!

kandersen replied:
Can you help me understand why it's not available? Seems to me it ought to be given the current build configuration.

udalov replied:
It's declared in the module `:compiler:ir.backend.common`, which `:compiler:ir.psi2ir` doesn't depend on

udalov commented on compiler/ir/ir.psi2ir/src/org/jetbrains/kotlin/psi2ir/generators/FragmentFunctionGenerator.kt:
```diff
@@ -0,0 +1,81 @@
+            irFunction.body = createBodyGenerator(irFunction.symbol).generateBody()
+        }
+
+    @OptIn(ObsoleteDescriptorBasedAPI::class)
```
`@OptIn(ObsoleteDescriptorBasedAPI::class)` has no effect in psi2ir because this opt-in is provided at the module level (in `build.gradle.kts`), so let's remove all of these annotations

kandersen replied:
Done.

udalov commented on compiler/ir/ir.psi2ir/src/org/jetbrains/kotlin/psi2ir/generators/FragmentFunctionGenerator.kt:
```diff
@@ -0,0 +1,81 @@
+            descriptor, descriptor.type.toIrType(),
+            (descriptor as? ValueParameterDescriptor)?.varargElementType?.toIrType(),
+            name,
+            isAssignable = true
```
Could you clarify why it should be assignable? Maybe it deserves a comment with an explanation...

kandersen replied:
It's needed for the approach that the evaluator takes to modifications to captured variables. Captures are passed as parameters to the fragment code, modifications modify the _parameters_ directly, and the evaluator then reads back the new value from the stack frame of the interpreter/VM running the fragment, which is then written back to the original captured variable in the JVM being debugged.

udalov replied:
Thanks, got it. It would be great to add such explanation as a comment.

kandersen replied:
Done.

udalov commented on compiler/ir/ir.psi2ir/src/org/jetbrains/kotlin/psi2ir/generators/GeneratorContext.kt:
```diff
@@ -31,7 +32,8 @@ class GeneratorContext private constructor(
     override val irBuiltIns: IrBuiltIns,
-    internal val callToSubstitutedDescriptorMap: MutableMap<IrDeclarationReference, CallableDescriptor>
+    internal val callToSubstitutedDescriptorMap: MutableMap<IrDeclarationReference, CallableDescriptor>,
+    internal val capturedDescriptorToFragmentParameterMap: MutableMap<DeclarationDescriptor, IrValueParameterSymbol>
```
Let's slightly encapsulate this in a new class (`FragmentContext` or something like that?).

Among increasing readability a bit, it'll allow to make the map nullable and mutable, and initialized to null at the beginning, so that we could completely skip hash table lookup in `CallGenerator.generateValueReference` in the non-fragment case.

kandersen replied:
Done. Let me know if my take on this is what you had in mind. 

I don't see the need for the _map_ to be nullable, but I guess it's just a matter of where you put the `?`: on the additional context or the map :)

kandersen replied:
Will let you know when I am done with this first pass, so you can address all this at once.

udalov replied:
👍

udalov commented on compiler/ir/serialization.common/src/org/jetbrains/kotlin/backend/common/serialization/DescriptorByIdSignatureFinder.kt:
```diff
@@ -22,7 +22,9 @@ class DescriptorByIdSignatureFinder(
     init {
-        assert(lookupMode != LookupMode.MODULE_ONLY || moduleDescriptor is ModuleDescriptorImpl)
+        assert(lookupMode != LookupMode.MODULE_ONLY || moduleDescriptor is ModuleDescriptorImpl) {
+            "DescriptorByIdSignatureFinder.init"
```
Probably this was for debug and is no longer needed?

kandersen replied:
Yes, exactly. I cannot recall the specific circumstances, but some part of the test framework/error reporting stack swallowed this origin and just reported a non-descript error without this precise location.

Can we come up with something more fitting than `"DescriptorByIdSignatureFinder.init"` for the failing case, I think it will improve debugability.

udalov replied:
Sounds weird. :) In any case, let's either drop it, or replace with some meaningful message, e.g. `"Incorrect lookup mode $lookupMode for $moduleDescriptor"`

kandersen replied:
Done.

udalov commented on compiler/ir/backend.jvm/entrypoint/src/org/jetbrains/kotlin/backend/jvm/JvmIrFragmentCodegenFactory.kt:
```diff
@@ -0,0 +1,225 @@
+                val fileClassInfo = getFileClassInfoNoResolve(containingFile)
+                val fileClassName = JvmClassName.byFqNameWithoutInnerClasses(fileClassInfo.fileClassFqName)
+                val facadeClassName = JvmClassName.byFqNameWithoutInnerClasses(fileClassInfo.facadeClassFqName)
+                return JvmPackagePartSource(fileClassName, facadeClassName, ProtoBuf.Package.getDefaultInstance(), object : NameResolver {
```
Maybe create another `DeserializedContainerSource` implementation instead?

If there's some code in the backend which expects it to be exactly `JvmPackagePartSource`, I suppose we should introduce an intermediate interface with the needed parts between `JvmPackagePartSource` and `DeserializedContainerSource`, use it in that backend code, and implement it here

kandersen replied:
Did a take on this, let me know if I got your concerns right.

I'll admit I don't understand the full importance or mechanism of `DeserializedContainerSource`, but I introduced an interface for the backend (in `DeclarationStubGenerator`) to depend on, rather than explicitly on `JvmPackagePartSource`.

Then, I introduced a new implementation of `DeserializedContainerSource` letting me move some of this code into there. 

I added a bit of explanation for this to the code as well.

udalov replied:
Yes, this is exactly what I proposed, thanks!

udalov reviewed (changes requested):
Sorry for the delay, Kristoffer. I've left some comments

kandersen commented:
> Sorry for the delay, Kristoffer. I've left some comments

No worries, Alex, thank you for your time! I will address your comments asap.

kandersen commented on compiler/ir/ir.psi2ir/src/org/jetbrains/kotlin/psi2ir/generators/EvaluatorFragmentInfo.kt:
```diff
@@ -0,0 +1,26 @@
+ *  This data structure contains "synthesized" descriptors for that class,
+ *  method and parameter lay-out.
+ */
+class EvaluatorFragmentInfo(
```
Done.

kandersen commented on compiler/ir/ir.psi2ir/src/org/jetbrains/kotlin/psi2ir/generators/FragmentModuleGenerator.kt:
```diff
@@ -0,0 +1,60 @@
+        val fileEntry = PsiIrFileEntry(ktFile)
+        val packageFragmentDescriptor = context.moduleDescriptor.findPackageFragmentForFile(ktFile)!!
+        return IrFileImpl(fileEntry, packageFragmentDescriptor).apply {
+            metadata = DescriptorMetadataSource.File(CodegenUtil.getMemberDescriptorsToGenerate(ktFile, context.bindingContext))
```
Done.

kandersen commented on compiler/ir/ir.psi2ir/src/org/jetbrains/kotlin/psi2ir/Psi2IrTranslator.kt:
```diff
@@ -94,4 +92,32 @@ class Psi2IrTranslator(
         return irModule
     }
+
+    // TODO - De-duplicate this code from generateModuleFragment
```
Done.

kandersen commented:
@udalov I pushed work addressing all review comments save for the deduplication of `JvmIrFragmentCodegenFactory`.

Do go ahead and take a look, in particular my solutions to your concerns on `DeserializedContainerSource` and `GeneratorContext`.

kandersen commented on compiler/ir/backend.jvm/entrypoint/src/org/jetbrains/kotlin/backend/jvm/JvmIrFragmentCodegenFactory.kt:
```diff
@@ -0,0 +1,225 @@
+import org.jetbrains.kotlin.resolve.source.PsiSourceFile
+import org.jetbrains.kotlin.serialization.deserialization.descriptors.DeserializedContainerSource
+
+// Copied and modified from JvmIrCodegenFactory. The vision is that, as we
```
Done. Used arguments rather than callbacks. 

If you prefer call-backs I can rework it, but you are right that the actual differences are very small.

kandersen replied:
The 'call site' is here:

https://github.com/JetBrains/intellij-community/pull/1666/files#diff-6996bd288d755d156712e23e645a5459ac32d95fd09539fce543335ef0b1575aR137-R170

Possibly that needs to be cleaned a little. I guess I should lift it into a method for clarity. Let me know what you think of the approach, before I go any further.

udalov replied:
Yeah, looks great now and exactly what I had in mind

kandersen commented on compiler/ir/ir.psi2ir/src/org/jetbrains/kotlin/psi2ir/generators/fragments/FragmentCompilerSymbolTableDecorator.kt:
```diff
@@ -0,0 +1,59 @@
+
+    override fun referenceValue(value: ValueDescriptor): IrValueSymbol {
+        val parameterPosition =
+            fragmentInfo.parameters.map { it }.indexOf(value)
```
Done.

kandersen commented on compiler/ir/ir.psi2ir/src/org/jetbrains/kotlin/psi2ir/generators/fragments/FragmentCompilerSymbolTableDecorator.kt:
```diff
@@ -0,0 +1,59 @@
+            is ImplicitClassReceiver -> { targetDescriptor: DeclarationDescriptor ->
+                receiverValue.classDescriptor == targetDescriptor.original
+            }
+            else -> TODO("Unimplemented/unhandled/...unreachable?")
```
Done.

kandersen commented on compiler/ir/ir.psi2ir/src/org/jetbrains/kotlin/psi2ir/generators/fragments/FragmentFunctionGenerator.kt:
```diff
@@ -0,0 +1,86 @@
+            irFunction.body = createBodyGenerator(irFunction.symbol).generateBody()
+        }
+
+    private fun declareFragmentFunctionInner(descriptor: FunctionDescriptor, origin: IrDeclarationOrigin): IrSimpleFunction =
```
Done. Took it a _little_ further :)

kandersen commented on compiler/ir/ir.psi2ir/src/org/jetbrains/kotlin/psi2ir/generators/fragments/FragmentFunctionGenerator.kt:
```diff
@@ -0,0 +1,86 @@
+            UNDEFINED_OFFSET,
+            UNDEFINED_OFFSET,
+            IrDeclarationOrigin.DEFINED,
+            descriptor, descriptor.type.toIrType(),
```
Done. Odd intellij doesn't have an opinion on this.

kandersen commented on compiler/ir/ir.psi2ir/src/org/jetbrains/kotlin/psi2ir/generators/fragments/FragmentDeclarationGenerator.kt:
```diff
@@ -0,0 +1,56 @@
+    }
+}
+
+abstract class FragmentDeclarationGeneratorExtension(val declarationGenerator: FragmentDeclarationGenerator) : Generator {
```
Thanks for pointing this out! The complication is a remnant from my very first efforts in getting this working where I copied the approach from the existing psi2ir generators.

Now that it has settled, stepping back and simplifying is possible. I think this is much cleaner. The module and declaration generator could be similarly combined, but for organization purposes, 2 classes feel OK.

kandersen commented:
I am keeping the 3 "Code Review" commits separate for now - I think that will make it easier to review, but will squash them before merging.

udalov commented on compiler/ir/backend.jvm/entrypoint/src/org/jetbrains/kotlin/backend/jvm/FacadeClassSourceShimForFragmentCompilation.kt:
```diff
@@ -0,0 +1,41 @@
+
+    private val fileClassInfo = getFileClassInfoNoResolve(containingFile.psiFile as KtFile)
+
+    override val incompatibility: IncompatibleVersionErrorData<*>?
```
I'd actually implement all of this since it seems trivial :)
```
    override val incompatibility: IncompatibleVersionErrorData<*>?
        get() = null
    override val isPreReleaseInvisible: Boolean
        get() = false
    override val abiStability: DeserializedContainerAbiStability
        get() = DeserializedContainerAbiStability.STABLE
    override val presentableString: String
        get() = "Fragment for $containingFile"
```

kandersen replied:
Done.

udalov commented on compiler/ir/backend.jvm/entrypoint/src/org/jetbrains/kotlin/backend/jvm/FacadeClassSourceShimForFragmentCompilation.kt:
```diff
@@ -0,0 +1,41 @@
+import org.jetbrains.kotlin.serialization.deserialization.descriptors.DeserializedContainerAbiStability
+import org.jetbrains.kotlin.serialization.deserialization.descriptors.DeserializedContainerSource
+
+class FacadeClassSourceShimForFragmentCompilation(private val containingFile: PsiSourceFile) :
```
How come this class is not used right now?

kandersen replied:
I "see" this (`FacadeClassSourceShim ForFragmentCompilation`) and `FragmentCompilerSymbolTableDecorator` as the only two classes that are used nowhere else but from the plug-in.

I am running the "combined" project set-up, so my IDEA can "see" the uses from the plug-in and does not complain.

No usages planned from the main repo.

udalov replied:
Yeah, most of the compiler team uses the single project, so these will be marked as unused.

But it's not a big deal. The build on the build server will fail anyway if someone accidentally removes these declarations.

udalov commented on compiler/ir/backend.jvm/entrypoint/src/org/jetbrains/kotlin/backend/jvm/JvmIrCodegenFactory.kt:
```diff
@@ -71,7 +74,12 @@ open class JvmIrCodegenFactory(
+            state.bindingContext,
+            symbolTable,
+            jvmGeneratorExtensions,
+            fragmentContext = evaluatorFragmentInfoForPsi2Ir?.let { FragmentContext() })
```
```suggestion
            fragmentContext = if (evaluatorFragmentInfoForPsi2Ir != null) FragmentContext() else null
        )
```

kandersen replied:
Done.

udalov commented on compiler/ir/ir.psi2ir/src/org/jetbrains/kotlin/psi2ir/generators/fragments/EvaluatorFragmentInfo.kt:
```diff
@@ -0,0 +1,26 @@
+class EvaluatorFragmentInfo(
+    val classDescriptor: ClassDescriptor,
+    val methodDescriptor: FunctionDescriptor,
+    val parameters: List<DeclarationDescriptor>
```
Can this be `List<ParameterDescriptor>`? (common superinterface for `ValueParameterDescriptor` and `ReceiverParameterDescriptor`)

kandersen replied:
Mmmm, not _immediately_. The choice of type is inherited from the design of the `IParameter` in CodeFragmentCodegen, though I am unclear on whether their uses are always pointing to ParameterDescriptors.

Would you be happy with postponing this exploration for later? If not, I'd be happy to investigate now.

udalov replied:
Yeah, it's not a big deal, I was just curious, thanks

udalov commented on compiler/ir/ir.psi2ir/src/org/jetbrains/kotlin/psi2ir/generators/fragments/FragmentCompilerSymbolTableDecorator.kt:
```diff
@@ -0,0 +1,60 @@
+class FragmentCompilerSymbolTableDecorator(
+    signatureComposer: IdSignatureComposer,
+    irFactory: IrFactory,
+    nameProvider: NameProvider = NameProvider.DEFAULT,
```
I usually prefer to move parameters with default values to the end, to avoid requiring named arguments for subsequent parameters at the call site

kandersen replied:
Done.

udalov commented on compiler/ir/ir.psi2ir/src/org/jetbrains/kotlin/psi2ir/generators/fragments/FragmentModuleGenerator.kt:
```diff
@@ -0,0 +1,49 @@
+    override fun generateModuleFragment(
+        ktFiles: Collection<KtFile>,
+    ): IrModuleFragment {
+        assert(ktFiles.size == 1 && ktFiles.single() is KtBlockCodeFragment) {
```
Usually we avoid such double checking with
```
        val ktBlockCodeFragment = ktFiles.singleOrNull() as? KtBlockCodeFragment
            ?: error("Multiple fragments in one compilation not understood and implemented yet")
```

kandersen replied:
Done.

udalov commented on compiler/ir/ir.psi2ir/src/org/jetbrains/kotlin/psi2ir/Psi2IrTranslator.kt:
```diff
@@ -72,9 +77,13 @@ class Psi2IrTranslator(
     ): IrModuleFragment {
-        val moduleGenerator = ModuleGenerator(context, expectDescriptorToSymbol)
+        val moduleGenerator = fragmentInfo?.let {
+            FragmentModuleGenerator(context, fragmentInfo)
```
(extremely minor, but since you're using `let`...)
```suggestion
            FragmentModuleGenerator(context, it)
```

kandersen replied:
Done.

udalov commented on compiler/ir/ir.tree/src/org/jetbrains/kotlin/ir/util/StubGeneratorExtensions.kt:
```diff
@@ -26,6 +26,8 @@ open class StubGeneratorExtensions {
         stubGenerator: DeclarationStubGenerator,
     ): IrClass? = null
 
+    open fun getContainerSource(descriptor: DeclarationDescriptor): DeserializedContainerSource? = null
```
Is this method needed now, or can its usage be reverted to how it was before?

kandersen replied:
Makes perfect sense!

Done.

udalov reviewed (approved):
Looks great now! Could you please squash and rebase on master, and fix remaining minor comments?

kandersen commented:
Thanks for responding. Conflict resolution and squash incoming :)

kandersen commented:
Rebased and squashed, @udalov !

udalov commented:
Thank you. Great job! Merged to master.

kandersen commented:
Awesome, thank you! Stay tuned for more ;)
